src/main/java/org/protege/owlapi/concurrent/WriteSafeOWLOntologyImpl.java

```diff
@@ -1318,4 +1297,861 @@ public boolean isTopEntity() {
         }
     } 
     
+    @Override
+    public void accept(OWLNamedObjectVisitor visitor) {
+        readLock.lock();
+        try {
+            delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public <O> O accept(OWLNamedObjectVisitorEx<O> visitor) {
+        readLock.lock();
+        try {
+            return delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public void setOWLOntologyManager(OWLOntologyManager saveOntology) {
+        writeLock.lock();
+        try {
+            delegate.setOWLOntologyManager(saveOntology);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void saveOntology() throws OWLOntologyStorageException {
+        writeLock.lock();
```

Comment:
On 04/07/2015 03:19 PM, Matthew Horridge wrote:
>
> In 
> src/main/java/org/protege/owlapi/concurrent/WriteSafeOWLOntologyImpl.java 
> <https://github.com/protegeproject/protege-owlapi/pull/3#discussion_r27926869>:
>
> > +        }
> > +    }
> > +
> > +    @Override
> > +    public void setOWLOntologyManager(OWLOntologyManager saveOntology) {
> > +        writeLock.lock();
> > +        try {
> > +            delegate.setOWLOntologyManager(saveOntology);
> > +        } finally {
> > +            writeLock.unlock();
> > +        }
> > +    }
> > +
> > +    @Override
> > +    public void saveOntology() throws OWLOntologyStorageException {
> > +        writeLock.lock();
>
> If a read lock is locked then not changes can be applied. The 
> ReadWriteLock allows for multiple concurrent readers but only a single 
> writer. The write lock is only for things that change the actual 
> ontology. Well, that how I think it /should/ work!
>

I agree with your original assessment - we should use read locks on the 
save methods.  It has been a while but I don't an obvious motivation for 
using a write lock.  The difference would be that readers should not 
have to wait while a save is in progress.

> I think all of this needs starting afresh. I don't feel confident that 
> ProtegeOWLOntologyManager will work after this.
>

When I looked at this before, the reader-writer locks were a thread-safe 
approach for the OWL api that required minimal assumptions on the OWL 
api and additionally provided for atomic update operations.  My 
recollection was that the only thing that the reader-writer locks 
required was that

  * read operations on the OWL api should be thread safe.  This wasn't
    always true.
  * read operations did not return sets that changed as the OWL ontology
    changed.  This also wasn't always true and there was some discussion
    about what made sense. In the end the OWL api made copies of sets
    before returning them.  How does the new OWL api behave?

The disadvantage of the reader-writer approach is that when a writer is 
modifying the ontology, the readers are all locked out. This may have 
implications for performance.

The advantage of the reader-writer approach is that writes to an 
ontology are atomic which was not true for the other solutions (the 
various OWL api internals methods) that existed.

I studied the OWL api's methods for thread-safety and writes were not 
atomic. It was possible for separate threads intermediate states while a 
write was in progress.  Some data structures would be updated before 
others and it is possible for parallel reading threads to make reads 
where some of the data structures had been updated but others had not.  
While such thread-safety may work for some applications, it is very 
complex from a callers because there is no simple description of what an 
ontology update would do.  The much more conservative reader-writer 
approach had a very simple semantics - a write is atomic and there are 
no intermediate states.

I was able to demonstrate these intermediate states in a debugger (with 
careful stepping) but I never wrote a program that would give some 
insight into the consequences of the presence of the half written 
ontology write states.

I haven't looked at the new version of the OWL api and I don't know if 
things are significantly different now.


> Oh, and there's way too many different things going on in this library 
> :( Might separate out the concurrent stuff.
>

Yes that is probably true.  This api was the dumping ground of 
everything that did not have a place.  At one time I had hoped that the 
reader-writer locks would find a home in the OWL api because I thought 
that it was a simpler model than other solutions that existed at the time.

-Timothy


> btw are the streams in v5 streams on copies of collections?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/protegeproject/protege-owlapi/pull/3/files#r27926869>.
>